src/lib/stream/cc-stream.js

```diff
@@ -0,0 +1,461 @@
+        return;
+      }
+      default: {
+        if (message.id?.length > 0) {
+          this.#lastReceivedMessageId = message.id;
+        }
+
+        this.#debugLog('Emitting custom event', { event: message.event, hasData: !!message.data });
+        this.#emit(message.event, message.data);
+      }
+    }
+  }
+
+  #onClose() {
+    if (this.#state === 'closed' || this.#state === 'paused') {
+      this.#debugLog('Close event ignored', { currentState: this.#state });
+      return;
+    }
+    this.#debugLog('Stream closed unexpectedly by server', {
+      state: this.#state,
+      lastMessageId: this.#lastReceivedMessageId,
+      retryCount: this.#retryCount,
+    });
+    this.#onError(new CcClientError('Server closed the response without a END_OF_STREAM event', 'SSE_SERVER_ERROR'));
+  }
+
+  /**
+   * when the SSE has an error
+   * @param {CcClientError} error
+   */
+  #onError(error) {
+    if (this.#state === 'closed' || this.#state === 'paused') {
+      this.#debugLog('Error event ignored', { currentState: this.#state, error });
+      return;
+    }
+
+    this.#debugLog('Stream error occurred', error);
+
+    this.#cleanup();
+
+    const wrappedError = isNetworkError(error)
+      ? new CcClientError('Failed to establish/maintain the connection with the server', 'SSE_SERVER_ERROR', error)
+      : error;
+
+    const canRetry = this.#canRetry(wrappedError);
+    this.#debugLog('Retry decision', {
+      canRetry,
+      retryCount: this.#retryCount,
+      maxRetry: this.#config.retry?.maxRetryCount,
+    });
+
+    if (canRetry) {
+      this.#state = 'paused';
+      this.#emit('error', wrappedError);
+
+      this.#retryCount++;
+      const exponentialBackoffDelay =
+        this.#config.retry.initRetryTimeout * this.#config.retry.backoffFactor ** this.#retryCount;
+
+      this.#debugLog('Scheduling retry', {
+        retryCount: this.#retryCount,
+        delayMs: exponentialBackoffDelay,
+        backoffFactor: this.#config.retry.backoffFactor,
+      });
+
+      this.#retryTimeoutId = setTimeout(this.#start.bind(this), exponentialBackoffDelay);
+    } else {
+      this.#debugLog('Stream permanently closed due to error', {
+        finalRetryCount: this.#retryCount,
+        error: wrappedError,
+      });
+      this.#state = 'closed';
+      this.#closeDeferred.reject(wrappedError);
+    }
+  }
+
+  /**
+   * @param {any} error
+   * @returns {boolean}
+   */
+  #canRetry(error) {
+    if (this.#config.retry == null) {
+      return false;
+    }
+
+    if (this.#retryCount >= this.#config.retry.maxRetryCount) {
+      return false;
+    }
+
+    if (error == null) {
+      return true;
+    }
+
+    if (error instanceof CcHttpError) {
+      return error.statusCode >= 500;
+    }
+
+    return error instanceof CcClientError;
+  }
+
+  /**
+   * Construct and dispatch an event
+   *
+   * @param {string} type
+   * @param {any} [data]
+   * @returns {boolean}
+   */
+  #emit(type, data) {
+    // todo; try catch
```

Comment:
question: is it for later or you forgot?

Reply:
I forgot. thanks